yksgi/src/lib.rs

```diff
@@ -149,6 +168,271 @@ impl SGInterp {
+    // `memcpy`ed to the actual stack by ykcapi::__ykrt_reconstruct_frames.
+    pub fn reconstruct_frames(&self, btmframeaddr: *mut c_void) -> *const c_void {
+        // Load the stackmap and text sections from the binary. The former we need to parse in the
+        // stackmaps. The latter is required to disassemble the binary in order to distinguish
+        // branches from calls.
+        // FIXME: Don't use current_exe.
+        let pathb = env::current_exe().unwrap();
+        let file = fs::File::open(&pathb.as_path()).unwrap();
+        let exemmap = unsafe { memmap2::Mmap::map(&file).unwrap() };
+        let object = object::File::parse(&*exemmap).unwrap();
+        let sec = object.section_by_name(".llvm_stackmaps").unwrap();
+        let text = object.section_by_name(".text").unwrap();
+
+        // Create a capstone dissassembler.
+        // so we can figure out the size of the instruction we
+        // return too in order to adjust the return address in our reconstructed stackmap.
+        let cs = Capstone::new()
+        .x86()
+        .mode(arch::x86::ArchMode::Mode64)
+        .syntax(arch::x86::ArchSyntax::Att)
+        .detail(true)
+        .build()
+        .expect("Failed to create Capstone object");
+
+        // Vec holding currently active register values.
+        let mut registers = vec![0; 16];
+
+        // Parse the stackmap.
+        let slice = unsafe { slice::from_raw_parts(sec.address() as *mut u8, usize::try_from(sec.size()).unwrap()) };
+        let smentries = StackMapParser::get_entries(slice);
+
+        // The final size of the memory we need to allocate. Inititialised with space to store
+        // registers for register recovery.
+        let mut memsize: usize = 15 * 8;
+        // Vec to collect stackmaps for each frame.
+        let mut smaps = Vec::new();
+
+        // Collect stackmaps for each frame and calculate the final size of memory required to
+        // store the reconstructed stack.
+        for (i, frame) in self.frames.iter().enumerate() {
+            // Get stackmap ID for the current frame's pc.
+            let smcall = unsafe { get_stackmap_call(frame.pc) };
+            let smid = unsafe { LLVMConstIntGetZExtValue(smcall.get_operand(0).get()) };
+            // Find prologue info and stackmap record for this frame.
+            let mut pinfo = None;
+            let mut rec = None;
+            // Iterate over function entries to find the correct record and relevant prologue info.
+            for entry in &smentries {
+                for r in &entry.records {
+                    if r.id == smid {
+                        pinfo = Some(&entry.pinfo);
+                        rec = Some(r);
+                        break;
+                    }
+                }
+            };
+            let rec = rec.unwrap();
+            let pinfo = pinfo.unwrap();
+            if i > 0 {
+                // We don't need to include the bottom-most frame, which is the frame containing
```

Comment:
I don't understand how the comment relates to the `memsize +=`.

Reply:
Would it help if it was above the condition? The comment explains why we skip `i=0`.

Reply:
Fixed in [c45e5e6](https://github.com/ykjit/yk/pull/578/commits/c45e5e644b2bc6b43c9d456a8192a5cc3ddbaff8)

Reply:
Got you. So `rec.size` is the size of each stack frame. In order to calculate how much memory we need to allocate for out new stack, we add the frame sizes of all frames together. However, the stackmap (which we iterate over) includes the bottom frame (the caller of the control point). But since that frame already exists, we don't need to allocate space for it. Does that clear things up?

Reply:
See if [1f2afe8](https://github.com/ykjit/yk/pull/578/commits/1f2afe8397349af56993f7de0c1e1152aa212eac) works better for you.

Reply:
Got it!